File: 2707-Extra Characters in a String.py
class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        
# bottom-up
        N = len(s)
        dp = [0] * (N+1)
        dictionary = set(dictionary)
        
        for i in range(N):
            for j in range(i+1):
                if s[j:i+1] in dictionary:
                    dp[i+1] = max(dp[i+1], dp[j] + (i - j + 1))
                dp[i+1] = max(dp[i+1], dp[i])
        
        return len(s) - dp[-1]
    
# top-down
        dictionary = set(dictionary)
        N = len(s)
        
        @cache
        def dfs(ind):
            if ind < 0:
                return 0
            res = dfs(ind-1)
            for j in range(ind+1):
                if s[j:ind+1] in dictionary:
                    res = max(res, dfs(j-1) + (ind-j+1))
            return res
        
        return len(s) - dfs(N-1)

# bottom-up2 
        dictionary = set(dictionary)
        N = len(s)
        dp = [0]*(N+1)
        
        for start in range(N-1, -1, -1):
            dp[start] = 1 + dp[start+1]
            for end in range(start, N):
                if s[start:end+1] in dictionary:
                    dp[start] = min(dp[end+1], dp[start])
        
        return dp[0]

# DP + Trie
class TrieNode:
    def __init__(self):
        self.children = dict()
        self.isWord = False

class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        N = len(s)
        
        def buildTrie():
            root = TrieNode()
            for word in dictionary:
                node = root
                for w in word:
                    if w not in node.children:
                        node.children[w] = TrieNode()
                    node = node.children[w]
                node.isWord = True
            return root
        
        root = buildTrie()
        
        @cache
        def dfs(start):
            if start == N:
                return 0
            
            res = 1 + dfs(start+1)
            node = root
            
            for end in range(start, N):
                if s[end] in node.children:
                    node = node.children[s[end]]
                    if node.isWord:
                        res = min(res, dfs(end+1))
                else:
                    break
            return res
            
        return dfs(0)

File: C++/2707-Extra Characters in a String.cpp
class TrieNode {
public:
    unordered_map<char, TrieNode*> children;
    bool isEnd;
    TrieNode (): isEnd(false) {
    }
};

class Solution {
public:
    void dfs(int ind, string& s, TrieNode* root, vector<int>& dp) {
        if (!root->children.contains(s[ind])) return;
        int currIdx = ind;
        while (ind>=0) {
            if (root->children.contains(s[ind])) {
                root = root->children[s[ind]];
            } else {
                break;
            }
            if (root->isEnd) {
                dp[ind] = max(dp[ind], dp[currIdx+1]+currIdx-ind+1);
            }
            ind--;
        }
    }
    int minExtraChar(string s, vector<string>& dictionary) {
        TrieNode* root = new TrieNode();
        for (string d:dictionary) {
            TrieNode* node = root;
            for (int i=d.size()-1; i>=0; i--) {
                if (!node->children.contains(d[i])) {
                    node->children[d[i]] = new TrieNode();
                }
                node = node->children[d[i]];
            }
            node->isEnd = true;
        }
        int N = s.size();
        vector<int> dp(N+1, 0);  // dp[i]: number of used characters from index N-1 to i

        for (int i=N-1; i>=0; i--) {
            dfs(i, s, root, dp);
            dp[i] = max(dp[i], dp[i+1]);
        }


        return N - dp[0];
    }
};



// faster Trie solution
class Solution {
public:
    struct TrieNode {
        TrieNode* children[26];
        bool isEnd;
        explicit TrieNode():isEnd(false){
            std::fill(children, children +26 , nullptr);
        }
        virtual ~TrieNode(){ // this makes code even faster
            for(TrieNode* e : children) delete e;
        }
    };
    void dfs(int ind, string& s, TrieNode* root, vector<int>& dp) {
        if (root->children[s[ind]-'a'] == nullptr) return;
        int currIdx = ind;
        while (ind>=0) {
            if (root->children[s[ind]-'a'] != nullptr) {
                root = root->children[s[ind]-'a'];
            } else {
                break;
            }
            if (root->isEnd) {
                dp[ind] = max(dp[ind], dp[currIdx+1]+currIdx-ind+1);
            }
            ind--;
        }
    }
    int minExtraChar(string s, vector<string>& dictionary) {
        TrieNode* root = new TrieNode();
        for (string d:dictionary) {
            TrieNode* node = root;
            for (int i=d.size()-1; i>=0; i--) {
                if (node->children[d[i]-'a'] == nullptr) {
                    node->children[d[i]-'a'] = new TrieNode();
                }
                node = node->children[d[i]-'a'];
            }
            node->isEnd = true;
        }
        int N = s.size();
        vector<int> dp(N+1, 0);  // dp[i]: number of used characters from index N-1 to i

        for (int i=N-1; i>=0; i--) {
            dfs(i, s, root, dp);
            dp[i] = max(dp[i], dp[i+1]);
        }


        return N - dp[0];
    }
};